Rename index folder to index_uuid

areek opened the pull request:
Following up https://github.com/elastic/elasticsearch/pull/16217 , This PR uses `${data.paths}/nodes/{node.id}/indices/{index.uuid}` 
instead of `${data.paths}/nodes/{node.id}/indices/{index.name}` pattern to store index 
folder on disk.
This way we avoid collision between indices that are named the same (deleted and recreated).

Closes #13265
Closes #13264
Closes #14932
Closes #15853
Closes #14512

areek commented:
This is still a work in progress but would be good to have it reviewed.
IMO, we should rename pre-3.0 index folder when loading the global state for the first time in `GatewayMetaState`. This means once the indices are renamed they can't be read by 2.x. WDYT @bleskes @s1monw?

bleskes commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -100,16 +99,18 @@ public NodePath(Path path, Environment environment) throws IOException {
          */
-        public Path resolve(Index index) {
-            return indicesPath.resolve(index.getName());
+        public Path resolve(String indexUUID) {
```
can always use the Index class to refer to an index? we should get in this habit so we always have both name and uuid available.

s1monw replied:
++

areek replied:
fixed

bleskes commented:
@areek  you work in amazing speeds :) 

I scanned through this and I think this is the right direction. My only feedback is that we want to have the index folder be "INDEXNAME-UUID", for example if the index name is `test` and the uuid is `ax5fd` the folder will be `test-ax5fd` . The goal here is to make something both human readable and unique.

bleskes commented:
> we should rename pre-3.0 index folder when loading the global state for the first time in GatewayMetaState. 

Agreed - see 2.x's MultiDataPathUpgrader.upgradeMultiDataPath for an example (which I think you alread saw :) ). 

> This means once the indices are renamed they can't be read by 2.x. 

Correct, but Simon pointed out that this is easy to reverse by renaming the folders back. I **do** think we should double check how the dangling index logic in 2.x will behave with these renamed folders (to 2.x, it's weird that a folder name is different than the what the index state file in it says).

s1monw commented:
```I scanned through this and I think this is the right direction. My only feedback is that we want to have the index folder be "INDEXNAME-UUID", for example if the index name is test and the uuid is ax5fd the folder will be test-ax5fd . The goal here is to make something both human readable and unique.```

maybe bikeshedding but should it be `test_ax5fd`?

dakrone commented:
> My only feedback is that we want to have the index folder be "INDEXNAME-UUID", for example if the index name is test and the uuid is ax5fd the folder will be test-ax5fd . The goal here is to make something both human readable and unique.

Yes please, I was going to leave the same feedback, `${index-name}-${uuid}` would be great. Or with a `_` like Simon said.

bleskes commented:
> maybe bikeshedding but should it be test_ax5fd?
sure :)

s1monw commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -444,6 +444,15 @@ public boolean hasIndex(String index) {
         return indices.containsKey(index);
     }
 
+    public boolean hasIndexUUID(String indexUUID) {
```
this should really be `hasIndex(Index index)` no?

areek replied:
this has been removed

s1monw commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -100,16 +99,18 @@ public NodePath(Path path, Environment environment) throws IOException {
          */
         public Path resolve(ShardId shardId) {
-            return resolve(shardId.getIndex()).resolve(Integer.toString(shardId.id()));
+            return resolve(shardId.getIndex().getUUID()).resolve(Integer.toString(shardId.id()));
```
I think we should add a `Index#getPathIdentifier()` to have a single place to calculate it?

areek replied:
great idea, thanks for the suggestion!

s1monw commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -105,31 +105,31 @@ void findNewAndAddDanglingIndices(MetaData metaData) {
+        final Set<String> indicesUUIDs;
         try {
-            indices = nodeEnv.findAllIndices();
+            indicesUUIDs = nodeEnv.findAllIndices();
```
I think we need to fix this too that we don't read the index name from the directory name. We have to read some descriptor which I think is available everywhere now if not it's not an index. Then when we have that we can just use Index.java as a key

areek replied:
Now we load all the indices state files upfront and then add them to dangling indices if they are not present in the cluster state or are not identified as dangling already. This behaviour is different from before where we only used to load the state of indices that were relevant. We also don't try to rename the index according to the folder name, like before. Is this what you were getting at by "read some descriptor"?

areek commented:
Thanks for the feedback @bleskes, @s1monw and @dakrone! I have changed the index folder name to `${index-name}_${uuid}` as suggested.

Would like to get some thoughts on the state of `DanglingIndicesState` see https://github.com/elastic/elasticsearch/pull/16442#discussion_r51973523. 

> Agreed - see 2.x's MultiDataPathUpgrader.upgradeMultiDataPath for an example (which I think you alread saw :) ).

Thanks for the pointer @bleskes. I will start on the bwc after we are happy with the rename.

clintongormley commented:
I thought we were going to deal with some of the issues around using null bytes, control chars, and non-ascii chars in index names with this PR as well? (eg the pathIdentifier could replace any chars in the name that don't match `/\w|\-/`

Happy to leave that for a separate PR though.

bleskes commented:
> (eg the pathIdentifier could replace any chars in the name that don't match /\w|\-/

seems like an easy one to add here... I don't see why not. @areek any objections?

areek commented:
>  I thought we were going to deal with some of the issues around using null bytes, control chars, and non-ascii chars in index names with this PR as well? 

Do you mean validating the index name upon index creation or replacing illegal characters on upgrade? We should do both, IMO tighten index name validation and rename pre-3x indices if they don't contain anything other then alphanumeric or hyphen characters in this pr. one problem is we might have collisions in index names for old indices, if we replace illegal characters on upgrade, index metadata is still keyed by index name rather than the index object. Thoughts?

bleskes commented:
> Do you mean validating the index name upon index creation or replacing illegal characters on upgrade? 

I think the idea is to make sure that index names are not bound by the limitation of the file system. This will make us free to decided what we want to enforce from ES on the higher levels, without having to take this OS/FS specific limitations. Since we always add a unique identifier to the folder name, it's OK to just clean up the index name part and replace potential troublesome chars. I think a simple replace in the getPathIdentifier() utility method. Nothing more.

clintongormley commented:
Just a note: users will no longer be able to import dangling indices under a different name, just by renaming the index folder.  Not sure there is anything we can do about this short of providing a command line tool to rename the index.

bleskes commented:
> users will no longer be able to import dangling indices under a different name, just by renaming the index folder

Correct. I think we're good with that. I believe that the current plan is to rename this folder to the right name on node start, which is how we plan to deal with BWC. Correct @areek ?

Another dangling indices limitation is that we may not be able to import a (Correctly named) folder because the cluster already has another index of the same name.  We should decide how to deal with it. Options are:
1) Ignore and leave the folder as is (log warnings though).
2) Wait a bit (2h I believe was an option before), log warnings and eventually delete.
3) Delete immediately as it is clear the index was replaced.

I'm inclined to go with (1) . Ideas?

clintongormley commented:
I'm good with option 1

areek commented:
Now we upgrade old indices on startup (rename index folder from `index.name` to `{name}_{index.uuid}`). 

> Just a note: users will no longer be able to import dangling indices under a different name, just by renaming the index folder.

Right, but users can still import dangling indices with valid index folder names. Dangling indices are imported on the next cluster state update, ex: execute cluster reroute command. 

We ignore dangling indices that are:
  - invalid (no index state)
  - valid and have the same name as an existing index in the cluster
  - valid with an invalid folder name (we read the index state to log the right name)

Still a work in progress (mainly for tests) but reviews are appreciated :), core tests are passing, including bwc tests.

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,265 @@
+     * Upgrades all indices found under <code>nodeEnv</code>. Already upgraded indices are ignored.
+     */
+    public static void upgrade(final Settings settings, final NodeEnvironment nodeEnv) throws IOException {
+        XContentType format = XContentType.fromMediaTypeOrFormat(settings.get(MetaStateService.FORMAT_SETTING, "smile"));
```
looking at the MetaDataStateFormat code, we should be able to read the format from the file - something stinks here :) . Also (different change) I think we should remove that setting and always write smile (and read what ever we find).

areek replied:
> looking at the MetaDataStateFormat code, we should be able to read the format from the file

Do you mean that MetaDataStateFormat#loadLatestState reads the format from the file? I hardcoded the format to be `SMILE`

> Also (different change) I think we should remove that setting and always write smile (and read what ever we find).

I will open an issue for this.

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,265 @@
+        Files.createDirectories(indexPath);
+        try (SimpleFSDirectory directory = new SimpleFSDirectory(indexPath)) {
+            try (final Lock lock = directory.obtainLock(IndexWriter.WRITE_LOCK_NAME)) {
+                upgradeFiles(targetShardPath, indexPath, ShardPath.INDEX_FOLDER_NAME, shardPaths);
```
why do we need all these file copies? can't we just do a top level folder rename? consolidating to one folder is a 2.0 feature?

areek replied:
Simplified the upgrade by renaming the top level folder. Thanks for the suggestion.

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -83,10 +89,10 @@ public void processDanglingIndices(MetaData metaData) {
-                danglingIndices.remove(danglingIndex);
+        for (Index index : danglingIndices.keySet()) {
+            if (metaData.hasIndex(index.getName())) {
+                logger.debug("[{}] no longer dangling (created), removing from dangling list", index);
```
can we check the uuid and make sure it's the same and if not log a warning?

areek replied:
added

bleskes commented on core/src/main/java/org/elasticsearch/gateway/MetaStateService.java:
```diff
@@ -108,11 +104,36 @@ MetaData loadFullState() throws Exception {
+                if (indexFolderName.equals(indexPathId)) {
+                    indexMetaDataList.add(indexMetaData);
+                } else {
+                    logger.info("[{}] invalid folder name, rename to [{}], ignoring", indexFolderName, indexPathId);
```
this feels too lenient to me. At the very least we should add a parameter indicating whether it's OK to be lenient (dangling indices detection ) and be strict (throw an exception on node start up).

areek replied:
Now we throw an ISE on startup when an invalid index folder name is found

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -104,33 +110,27 @@ void findNewAndAddDanglingIndices(MetaData metaData) {
+                    indexMetaData.getIndex().getPathIdentifier());
+                newIndices.put(indexMetaData.getIndex(), indexMetaData);
+            } else {
+                logger.info("[{}] dangling index, exists on local file system and in cluster metadata, ignoring", indexMetaData.getIndex().getPathIdentifier());
```
I don't think this is possible any more? we added a filter on the path names?

bleskes commented:
@areek I left some initial feedback. Thx.

uboness commented:
While we're at it, I'd vote for completely removing the index names from the FS dir structure. In many cases, having these names as part of the dir names is a security concern. Instead, just use UUIDs. We can still provide a cli tool that helps map dirs to indices based on metadata files (for "offline" data dir "browsing")

areek commented:
Updated PR: 
 - Simplified bwc logic (rename top level index folder)
 - use index.uuid as the index folder name

Thanks @bleskes for the feedback, I addressed all the comments

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,121 @@
+        if (Files.exists(sourceDir)) {
+            logger.info("{} moving from [{}] to [{}]", index, sourceDir, targetDir);
+            Files.move(sourceDir, targetDir, StandardCopyOption.ATOMIC_MOVE);
+            logger.info("{} syncing directory [{}]", index, targetDir);
```
I think we can make this trace?

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,121 @@
+    void upgrade(final Index index, final Path sourceDir) throws IOException {
+        final Path targetDir = sourceDir.resolveSibling(index.getUUID());
+        if (Files.exists(sourceDir)) {
+            logger.info("{} moving from [{}] to [{}]", index, sourceDir, targetDir);
```
can we add something that will explain why we're doing this of the unordained user? something ala upgrading indexing folder to new naming conventions

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,121 @@
+     */
+    void upgrade(final Index index, final Path sourceDir) throws IOException {
+        final Path targetDir = sourceDir.resolveSibling(index.getUUID());
+        if (Files.exists(sourceDir)) {
```
can we add a comment into why need this check? (we already renamed it before)

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * Upgrades all indices found under <code>nodeEnv</code>. Already upgraded indices are ignored.
+     */
+    public static void upgrade(final Settings settings, final NodeEnvironment nodeEnv) throws IOException {
```
call it upgradeIndicesIfNeeded?

bleskes commented:
I think we can take an even simpler approach. Instead of retrieving all indices and then rename all folders of each index, we should rather work on the state file level. That means list all index state files and the go one by one and make sure their parent directory is what we want it to be. This approach will be more resilient to node crashes half way the rename. Currently if some of the folders are renamed and some are not the code will be confused.

Also - we should double check that shard replicas are well behaved (and add tests, including half way crashes and concurrent modifications :)).

areek commented:
Thanks @bleskes for the feedback! Updated the PR with the suggested approach (much simpler now). I added tests to ensure we behave as expected:
 - when we crash while upgrading
 - when multiple nodes try to rename folders concurrently in a shared FS (shadow replicas)

IMO, this is getting close, would appreciate a review :)

@dakrone would be awesome if you could take a look at the [test mocking upgrade on shared FS](https://github.com/areek/elasticsearch/commit/ad7df211cc595cc51a7e83234191744fc6e8e80b)

s1monw commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+        try {
+            Files.move(source, target, StandardCopyOption.ATOMIC_MOVE);
+            success = true;
+        } catch (NoSuchFileException ignored) {
```
this might also be a `FileNotFoundException`? ie `} catch (NoSuchFileException| | FileNotFoundException ignored) {`

s1monw commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+            assert ignored.getFile().equals(source.toAbsolutePath().toString()) : "must be same as the source";
+            // thrown when the source is non-existent because the folder was renamed
+            // by another node (shared FS) after we checked if the target exists
+            assert Files.exists(target) : "must be upgraded by another node";
```
wow this is scary as shit I guess this means we are restarting multiple nodes at the same time in a full cluster restart. I think we can't do this neither support it on a shared FS. If we run into this situation we should fail and not swallow exceptions IMO

s1monw replied:
lets document that in the migration guides

bleskes replied:
Agreed on not supporting it and using documentation instead. We should document that people should ideally start a single master node first - which has access to the data folders -, let it upgrade *all* the folders and then start up the rest of the cluster.

areek replied:
I removed support for this and documented it in the migration guide. Now we throw the exception and log an error message to retry upgrading when we detect `NoSuchFileException` or `FileNotFoundException`

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/ShardPath.java:
```diff
@@ -168,34 +161,6 @@ public static void deleteLeftoverShardDirectory(ESLogger logger, NodeEnvironment
 
-    /** Maps each path.data path to a "guess" of how many bytes the shards allocated to that path might additionally use over their
-     *  lifetime; we do this so a bunch of newly allocated shards won't just all go the path with the most free space at this moment. */
-    private static Map<Path,Long> getEstimatedReservedBytes(NodeEnvironment env, long avgShardSizeInBytes, Iterable<IndexShard> shards) throws IOException {
```
@mikemccand this is unused and get removed in 4d38856f7017275e326df52a44b90662b2f3da6a - was this a mistake or did you just not remove this method?

areek replied:
I removed this method intentionally as it is dead code

s1monw commented:
I love this change really nice! left some minor comments

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+    void upgrade(final String indexFolderName) throws IOException {
+        for (NodeEnvironment.NodePath nodePath : nodeEnv.nodePaths()) {
+            final Path indexFolderPath = nodePath.indicesPath.resolve(indexFolderName);
+            if (Files.exists(indexFolderPath) == false) {
```
this should happen right? we just got it for a dir list?

areek replied:
no this should never happen unless a folder has been manually deleted after we get `NodeEnvironment#availableIndexFolders`, removed the check

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+                    }
+                    upgrade(index, indexFolderPath);
+                } else {
+                    logger.info("[{}] no upgrade needed - already upgraded", indexFolderPath);
```
can we log this in debug? we're going to log it on every node start..

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+                    logger.info("[{}] no upgrade needed - already upgraded", indexFolderPath);
+                }
+            } else {
+                logger.info("[{}] no index state found - ignoring", indexFolderPath);
```
wondering if this should be a warn... it means we have an unknow folder in our universe?

bleskes replied:
it seems NodeEnv.availableIndexFolders checks for state files - we should really double check when this is possible and see whether we should rename it anyway (node crashed while not writing all state files) or throw hard errors..

areek replied:
thinking about this more, we should throw a hard exception when we find an unknown folder (has no state folder or has no valid state) during the upgrade. This way we ensure there are no unknown folders in the indices path once we successfully upgrade.

areek replied:
Now we just log it as a warn rather than throw a hard exception in this case, as we might have a node crash as it is [writing the index state](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java#L161). This may leave folders in indices folder with no valid index state. Currently, if a node crashed in such a state before being upgraded, the empty folder is left in the file system (we don't rename). If a node crashed in such a state after being upgraded, the upgrader just ignores it and let cluster state listener deal with it.

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+        return indexFolderName.equals(index.getUUID()) == false;
+    }
+
+    static MetaDataStateFormat<IndexMetaData> metaDataMetaDataStateFormat() {
```
can we name this readOnlyMetaDataMetaDataStateFormat

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+
+            @Override
+            public void toXContent(XContentBuilder builder, IndexMetaData state) throws IOException {
+                IndexMetaData.Builder.toXContent(state, builder, ToXContent.EMPTY_PARAMS);
```
throw an unsupported operaton?

areek replied:
done

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    static MetaDataStateFormat<IndexMetaData> metaDataMetaDataStateFormat() {
+        return new MetaDataStateFormat<IndexMetaData>(XContentType.SMILE, MetaStateService.INDEX_STATE_FILE_PREFIX) {
```
can we add a comment that this XContentType is not used? (and we should really remove the config option to change it :) - different change)

areek replied:
> can we add a comment that this XContentType is not used?

done

> and we should really remove the config option to change it :) - different change

I will open a PR for this

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,155 @@
+     */
+    void upgrade(final Index index, final Path source) throws IOException {
+        final Path target = source.resolveSibling(index.getUUID());
+        if (Files.exists(target)) {
```
can we only do this in the custom data folder case, before we call this method? this we can make this method as strict as it needs to be for the standard , local only code paths.

areek replied:
Thanks for the suggestion, now we check for target folder existence only for the custom data folder case

bleskes commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -646,10 +601,10 @@ private void decWaitCount() {
                 assert waitCount > 0 : "waitCount is " + waitCount + " but should be > 0";
                 --waitCount;
-                logger.trace("shard lock wait count for [{}] is now [{}]", lockKey, waitCount);
+                logger.trace("shard lock wait count for [{}] is now [{}]", shardId, waitCount);
```
shardIds already add the [] in their toString..

bleskes commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -646,10 +601,10 @@ private void decWaitCount() {
                 if (waitCount == 0) {
-                    logger.trace("last shard lock wait decremented, removing lock for [{}]", lockKey);
-                    InternalShardLock remove = shardLocks.remove(lockKey);
+                    logger.trace("last shard lock wait decremented, removing lock for [{}]", shardId);
```
same about  `[]`

bleskes commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -889,10 +868,10 @@ private Path resolveCustomLocation(IndexSettings indexSettings) {
      */
-    private Path resolveCustomLocation(IndexSettings indexSettings, final String indexName) {
-        return resolveCustomLocation(indexSettings).resolve(indexName);
+    private Path resolveCustomLocation(IndexSettings indexSettings, final Index index) {
```
IndexSettings already has the uuid? do we really need two params?

areek replied:
good catch, removed `index` as param

bleskes commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -928,22 +907,24 @@ private void assertCanWrite() throws IOException {
-                    tryWriteTempFile(translogDir);
-                    tryWriteTempFile(statePath);
-                    tryWriteTempFile(path);
+        for (String indexFolderName : this.availableIndexFolders()) {
```
this change has the down side that we don't try to write to folders that don't have any state file in them. I think we should?

areek replied:
Now `availableIndexFolders` doesn't filter out folders with no state file in them rather on upgrade we throw a hard exception whenever we find a folder without a valid state.

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -83,10 +89,15 @@ public void processDanglingIndices(MetaData metaData) {
+            final IndexMetaData indexMetaData = metaData.index(index);
+            if (indexMetaData != null && indexMetaData.getIndex().getName().equals(index.getName())) {
+                if (indexMetaData.getIndex().getUUID().equals(index.getUUID()) == false) {
+                    logger.warn("[{}] no longer dangling, index with same name but different uuid found", index);
```
"[{}] can not be imported as a dangling index, as there is already another index with the same name but a different uuid. local index will be ignored (but not deleted)"?

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -104,36 +115,28 @@ void findNewAndAddDanglingIndices(MetaData metaData) {
+            Map<Index, IndexMetaData> newIndices = new HashMap<>(indexMetaDataList.size());
+            indexMetaDataList.stream()
+                .filter(indexMetaData ->
+                    metaData.hasIndex(indexMetaData.getIndex().getName()) == false && danglingIndices.containsKey(indexMetaData.getIndex()) == false)
```
we should log a warn message here. I think it will all be cleaner to read without the filtering fancy pancing :)

areek replied:
now we log a warn message when we find a dangled index that has the same name as an existing index in cluster metadata.

bleskes commented on core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -84,6 +85,7 @@ public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateSer
             try {
                 ensureNoPre019State();
+                IndexFolderUpgrader.upgradeIndicesIfNeeded(settings, nodeEnv);
                 pre20Upgrade();
```
not sure about this - the pre20upgrade thing checks for old versions of state files and issues the proper message (via MetaDataIndexUpgradeService#checkSupportedVersion ). I think we should check for too old versions before starting to mess up with folders. People trying to go from a 1.x cluster to 5.0 should be rejected before we start messing up with folders.

bleskes replied:
and we should see what of pre20Upgrade is needed in 5.0 (different change)

areek replied:
> I think we should check for too old versions before starting to mess up with folders.

Agreed, now we upgrade index folders after we ensure version of indices are supported

bleskes commented on core/src/test/java/org/elasticsearch/cluster/DiskUsageTests.java:
```diff
@@ -92,22 +92,22 @@ public void testRandomDiskUsage() {
 
     public void testFillShardLevelInfo() {
-        final Index index = new Index("test", "_na_");
+        final Index index = new Index("test", "0xdeadbeef");
```
:)

bleskes commented on core/src/test/java/org/elasticsearch/cluster/allocation/ClusterRerouteIT.java:
```diff
@@ -226,9 +229,11 @@ private void rerouteWithAllocateLocalGateway(Settings commonSettings) throws Exc
         assertThat(state.getRoutingNodes().node(state.nodes().resolveNode(node_1).id()).get(0).state(), equalTo(ShardRoutingState.STARTED));
 
         client().prepareIndex("test", "type", "1").setSource("field", "value").setRefresh(true).execute().actionGet();
+        final ClusterStateResponse response = client().admin().cluster().prepareState().get();
```
Simon added a fancy resolveIndex method

bleskes commented on core/src/test/java/org/elasticsearch/gateway/MetaDataWriteDataNodesIT.java:
```diff
@@ -170,7 +171,9 @@ protected void assertIndexInMetaState(final String nodeName, final String indexN
     private boolean indexDirectoryExists(String nodeName, String indexName) {
         NodeEnvironment nodeEnv = ((InternalTestCluster) cluster()).getInstance(NodeEnvironment.class, nodeName);
-        for (Path path : nodeEnv.indexPaths(indexName)) {
+        ClusterService clusterService = ((InternalTestCluster) cluster()).getInstance(ClusterService.class, nodeName);
```
same comment about resolving. maybe easier to change the method to get an Index object as a param?

bleskes commented on core/src/test/java/org/elasticsearch/common/util/IndexFolderUpgraderTests.java:
```diff
@@ -0,0 +1,408 @@
+            int numTranslogFiles = randomIntBetween(1, 5);
+            IndexSettings indexSettings = new IndexSettings(indexState, nodeSettings);
+            writeIndex(nodeEnv, indexSettings, numIdxFiles, numTranslogFiles);
+            IndexFolderUpgrader helper = new IndexFolderUpgrader(settings, nodeEnv);
```
maybe add a method in IndexFolderUpgrader that does the rename and override it here to randomly throw exceptions? then we can iterated a couple of times, stop throwing exceptions, give it another go and see all is ok?

areek replied:
Thanks for the suggestion, in the test we randomly throw IOE in the first pass and check if upgrade on partially upgraded indices behave as expected

bleskes commented:
Thx @areek . I made another round and it looks really good. Left some comments, most importantly around the the custom path case and validating state versions in advance.

bleskes commented on core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java:
```diff
@@ -0,0 +1,153 @@
+                        // we rename index folder in custom path before renaming them in any node path
+                        // to have the index state under a not-yet-upgraded index folder, which we use to
+                        // continue renaming after a incomplete upgrade.
+                        final Path customLocationSource = nodeEnv.resolveBaseCustomLocation(indexSettings)
```
can we also check that this location exists? I'm not sure we can guarantee that the hasCustomDataPath is created... @dakrone - any opinions on this?

areek replied:
aah, so when we have the index state but there is no data yet?

dakrone replied:
Yeah, I think if the settings are there but the index has not been opened yet (don't ask, it's a weird but valid use case), the location that the settings point to may not exist

areek commented:
Thanks @bleskes and @s1monw for the feedback, I addressed all the comments!

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -104,36 +116,30 @@ void findNewAndAddDanglingIndices(MetaData metaData) {
+            final List<IndexMetaData> indexMetaDataList = metaStateService.loadIndicesStates(excludeIndexPathIds::contains);
+            Map<Index, IndexMetaData> newIndices = new HashMap<>(indexMetaDataList.size());
+            for (IndexMetaData indexMetaData : indexMetaDataList) {
+                if (metaData.hasIndex(indexMetaData.getIndex().getName()) == false && danglingIndices.containsKey(indexMetaData.getIndex()) == false) {
```
I'm confused by this danglingIndices.conainsKey() - isn't supposed to be filtered out by the excludeIndexPathIds?

areek replied:
This was left over, we don't need the danglingIndices.conainsKey() check anymore

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -104,36 +116,30 @@ void findNewAndAddDanglingIndices(MetaData metaData) {
+                    logger.info("[{}] dangling index, exists on local file system, but not in cluster metadata, auto import to cluster state",
+                        indexMetaData.getIndex());
+                    newIndices.put(indexMetaData.getIndex(), indexMetaData);
+                } else if (metaData.hasIndex(indexMetaData.getIndex().getName())) {
```
put this first?

bleskes commented:
Thx @areek . Left one question about hasCustomDataPath strictness. O.w. looks good to me.
